abn/metricstore/metric.go

```diff
@@ -0,0 +1,258 @@
+package metricstore
+
+import (
+	"context"
+	"errors"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/iter8-tools/iter8/base/log"
+	"github.com/iter8-tools/iter8/driver"
+	"sigs.k8s.io/yaml"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+const (
+	VERSIONS_DATA = "versionData.yaml"
+)
+
+type ApplicationData map[string]VersionData
+
+type VersionData struct {
+	Metrics             map[string]SummaryMetric `json:"metrics,omitempty" yaml:"metrics,omitempty"`
+	LastUpdateTimestamp time.Time                `json:"lastUpdateTimestamp" yaml:"lastUpdateTimestamp"`
+	History             []VersionEvent           `json:"history,omitempty" yaml:"history,omitempty"`
+}
+
+type VersionEvent struct {
+	Timestamp time.Time        `json:"tm" yaml:"tm"`
+	Event     VersionEventType `json:"ev" yaml:"ev"`
+	Track     string           `json:"trk,omitempty" yaml:"trk,omitempty"`
+}
+
+type VersionEventType string
+
+const (
+	VersionNewEvent           VersionEventType = "new"
+	VersionReadyEvent         VersionEventType = "ready"
+	VersionNoLongerReadyEvent VersionEventType = "notready"
+	VersionMapTrackEvent      VersionEventType = "track"
+	VersionUnmapTrackEvent    VersionEventType = "untrack"
+)
+
+// A MetricStore implemented as a Kubernetes secret
+type MetricStoreSecret struct {
+	// App name
+	App string
+	// Name of the ecret to use
+	Name string
+	// Namepsace of the secret
+	Namespace string
+	// client is kubernetes client used to read/write the secret
+	client kubernetes.Interface
+}
+
+// NewMetricStoreSecret creates a new MetricStoreSecret
+func NewMetricStoreSecret(app string, kd *driver.KubeDriver) (*MetricStoreSecret, error) {
+	if kd.Clientset == nil {
+		return nil, errors.New("a Kubernetes clientset is required to create MetricStoreSecret")
+	}
+
+	var ns, nm string
+	names := strings.Split(app, "/")
+	if len(names) > 1 {
+		ns, nm = names[0], names[1]
+	} else {
+		ns, nm = "default", names[0]
+	}
+
+	return &MetricStoreSecret{
+		App:       app,
+		Name:      nm,
+		Namespace: ns,
+		client:    kd.Clientset,
+	}, nil
+}
+
+type MetricStoreSecretCache struct {
```

Comment:
It would be more intuitive, if we get rid of the cache struct altogether ... and define its methods directly on the ApplicationData struct... 

This way, you have methods to read and write ApplicationData object into the secret, and you also have methods to manipulate the ApplicationData object itself.

Reply:
The primary motivator for this object was to cache the secret so that between the read and increment it could be written w/o rereading and parsing it again. On inspection, it seems pretty complicated. I agree that some simplification and clarification is a good idea.

Reply:
To achieve this caching, could we have an in-memory map as follows:

```go
type AppDataMap map[string]ApplicationData
var appDataMap *AppDataMap
```

`appDataMap` is initialized from secrets (if any) when the abn service initializes.

```go
appDataMap = initFromSecrets()
```

 Whenever there is a need to read, the read directly happens from `ApplicationData`. Below is an example (which can of course be suitably modified if need be)

```go
appDataMap[appName].getVersionForTrack(trackName) // this method does not involve read from secrets; so yes, caching.
```

Whenever there is a need to write, you can directly update the `AppDataMap`, which under the covers will also update the secret.

```go
appDataMap.write(appName, versionName, metricName, summaryMetric) // note that read is on AppData; write is on the data map
```

The above constructs remove the need for cache and simplify code; I wonder if they also eliminate the need for `MetricsSecretStore` data structure also.